src/databricks/labs/ucx/source_code/python/python_ast.py

```diff
@@ -48,88 +48,85 @@
 
 @dataclass(frozen=True)
 class MaybeTree:
-    tree: Tree | None
-    failure: Failure | None
-
-    def walk(self) -> Iterable[NodeNG]:
-        # mainly a helper method for unit testing
-        if self.tree is None:  # no cov
-            assert self.failure is not None
-            logger.warning(self.failure.message)
-            return []
-        return self.tree.walk()
+    """A :class:`Tree` or a :class:`Failure`.
+
+    The `MaybeTree` is designed to either contain a `Tree` OR a `Failure`,
+    never both or neither. Typically, a `Tree` is constructed using the
+    `MaybeTree` class method(s) that yields a `Failure` if the `Tree` could
+    NOT be constructed, otherwise it yields the `Tree`, resulting in code that
+    looks like:
```

Comment:
Please add a `__post_init__` method that enforces this contract: that one or other is present, and it's never neither or both.